demos/twa-offline-first/src/main/java/com/google/androidbrowserhelper/demos/twa_offline_first/OfflineFirstTWALauncherActivity.java

```diff
@@ -0,0 +1,62 @@
+// Copyright 2020 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.androidbrowserhelper.demos.twa_offline_first;
+
+import com.google.androidbrowserhelper.trusted.LauncherActivity;
+import android.content.Context;
+import android.net.ConnectivityManager;
+import android.net.NetworkInfo;
+import android.os.Bundle;
+import android.view.View;
+import android.widget.Button;
+
+public class OfflineFirstTWALauncherActivity extends LauncherActivity {
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        this.tryLaunchTwa();
+    }
+
+    @Override
+    protected boolean shouldLaunchImmediately() {
+        return false;
+    }
+
+    private void tryLaunchTwa() {
+        if (isOnline()) {
+            launchTwa();
+        } else {
+            renderOfflineFallback();
+        }
+    }
+
+    private void renderOfflineFallback() {
+        setContentView(R.layout.activity_offline_first_twa);
+
+        final Button retryBtn = this.findViewById(R.id.retry_btn);
+        retryBtn.setOnClickListener(new View.OnClickListener() {
+            public void onClick(View v) {
+                tryLaunchTwa();
```

Comment:
If the user is offline, and they click the button when they're still offline, we're going to be repeating all of this code even though there's no need to.

Maybe tryLaunchTwa should be something like:

```
if (isOnline()) {
  launchTwa();
} else if (!mOfflineFallbackRendered) {
  mOfflineFallbackRendered = true;
  renderOfflineFallback();
}
```

Or alternatively we just change this line to:

```
if (isOnline()) launchTwa();
```

Which is probably a bit simpler.